### DISCUSSION TITLE
Serving expired queries without prefetching

### DISCUSSION THREAD

**Discussion by agneevX** [0xERR0R/blocky#417] [General]
Hello!

I remember briefly discussing this in #290, but I'm still ambiguous as to whether this works as intended since this is not documented.

AdGuard Home 0.107.0 added a feature called Optimistic Caching where it serves expired records for an indefinite period of time while retrying upstream servers for fresh data.

If this feature exists, can it be documented and an option `serveExpired` added to disable or enable it?

**Comment by 0xERR0R** [comment]
Current implementation of prefetching renews cache entries automatically by async task. This task runs every 5 seconds. It can happen, that a client asks for the query and this query is in the cache, is already expired, but not prefetched yet (or prefetching is disabled). In this case the last cached value with ttl 0 will be returned.

Compared to the optimistic locking (AGH), I think it works similar, the difference is: blocky prefetches only often used entries and blocky prefetches automatically if the entry is expired and not on next request.

**Comment by agneevX** [comment]
> In this case the last cached value with ttl 0 will be returned.

Is there a time limit on how long stale results are returned? 
E.g. what happens when it can't reach upstreams?

**Reply by 0xERR0R** [reply]
Asynchronous task runs every 5 second -> if upstream fails, blocky will retry automatically (default 3 times with 1 second pause between).
